src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -25,39 +23,27 @@ internal abstract class AbstractRefactoringHelpersService<TExpressionSyntax, TAr
 
     public abstract bool IsBetweenTypeMembers(SourceText sourceText, SyntaxNode root, int position, [NotNullWhen(true)] out SyntaxNode? typeDeclaration);
 
-    public async Task<ImmutableArray<TSyntaxNode>> GetRelevantNodesAsync<TSyntaxNode>(
-        Document document, TextSpan selectionRaw, bool allowEmptyNodes, CancellationToken cancellationToken) where TSyntaxNode : SyntaxNode
+    private static void AddNode<TSyntaxNode>(bool allowEmptyNodes, ref TemporaryArray<TSyntaxNode> result, TSyntaxNode node) where TSyntaxNode : SyntaxNode
     {
-        using var _1 = ArrayBuilder<TSyntaxNode>.GetInstance(out var relevantNodesBuilder);
-        await AddRelevantNodesAsync(document, selectionRaw, relevantNodesBuilder, cancellationToken).ConfigureAwait(false);
-
-        if (allowEmptyNodes)
-            return relevantNodesBuilder.ToImmutableAndClear();
-
-        using var _2 = ArrayBuilder<TSyntaxNode>.GetInstance(out var nonEmptyNodes);
-        foreach (var node in relevantNodesBuilder)
```

Comment:
this loop went away, since we're only adding the nodes if they respect the allowEmptyNodes flag.